src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,263 @@
+     *
+     * @param key The resource key of the target object.
+     * @see #create(ResourceKey, ResourceLocation)
+     * @see #create(ResourceLocation, ResourceLocation)
+     * @see #create(ResourceKey)
+     */
+    protected DeferredHolder(ResourceKey<R> key) {
+        this.key = Objects.requireNonNull(key);
+        this.bind(false);
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public T value() {
+        bind(true);
+        if (this.holder == null) {
+            throw new NullPointerException("Trying to access unbound value: " + this.key);
+        }
+
+        return (T) this.holder.value();
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    public T get() {
+        return this.value();
+    }
+
+    /**
+     * @return An optional containing the target object, if {@link #isPresent()}, otherwise {@linkplain Optional#empty() an empty optional}.
+     */
+    public Optional<T> asOptional() {
+        return isPresent() ? Optional.of(value()) : Optional.empty();
+    }
+
+    /**
+     * The type of the registry is really <? super T> but this saves us some additional ugly casting and doesn't break anything.
+     *
+     * @return The registry that this DeferredHolder is pointing at, or null if it doesn't exist.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    protected Registry<R> getRegistry() {
+        return (Registry<R>) BuiltInRegistries.REGISTRY.get(this.key.registry());
+    }
+
+    /**
+     * Binds this DeferredHolder to the underlying registry and target object.<br> Has no effect if already bound.
+     *
+     * @param throwOnMissingRegistry If true, an exception will be thrown if the registry is absent.
+     * @throws IllegalStateException If throwOnMissingRegistry is true and the backing registry is unavailable.
+     */
+    protected void bind(boolean throwOnMissingRegistry) {
+        if (this.holder != null) return;
+
+        Registry<R> registry = getRegistry();
+        if (registry != null) {
+            this.holder = registry.getHolder(this.key).orElse(null);
+        } else if (throwOnMissingRegistry) {
+            throw new IllegalStateException("Registry not present for " + this + ": " + this.key.registry());
+        }
+    }
+
+    /**
+     * @return The ID of the object pointed to by this DeferredHolder.
```

Comment:
Simple getters should use inline `{@return}`. (Remove the final period/full stop, lowercase the first letter, and then surround with braces.)
```suggestion
     * {@return the ID of the object pointed to by this DeferredHolder}
```
Apply this suggestion to other simple getters.

Reply:
Most of the usages in this class would have a nested tag. They do not render correctly with IntelliJ, which is what most modders will use to browse the javadoc.

Reply:
I overhauled the javadoc of the simple getters, it should be correct now.